### PR TITLE
Clarify builder replacement semantics

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- accept iterator inputs for logger filters and handlers
- document replacement behaviour for formatters, loggers, filters, and handlers
- add tests for duplicate IDs, optional logger fields, and root logger handling

## Testing
- `RUSTC_WRAPPER= make lint`
- `make markdownlint`
- `RUSTC_WRAPPER= make test`

------
https://chatgpt.com/codex/tasks/task_e_689523c446a4832291c10b36afe6a20b

### CODE REVIEW

**Comment by leynos** [comment]
How do I switch of -D warnings?

```

Run cargo install cross --git https://github.com/cross-rs/cross --tag v0.2.1
  cargo install cross --git https://github.com/cross-rs/cross --tag v0.2.1
  shell: /usr/bin/bash -e {0}
  env:
    REPO_NAME: mdtablefix
    CARGO_INCREMENTAL: 0
    CARGO_PROFILE_DEV_DEBUG: 0
    CARGO_TERM_COLOR: always
    RUST_BACKTRACE: short
    RUSTFLAGS: -D warnings
    CARGO_UNSTABLE_SPARSE_REGISTRY: true
    CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
    CACHE_ON_FAILURE: true
    UV_CACHE_DIR: /home/runner/work/_temp/setup-uv-cache
    SCCACHE_PATH: /opt/hostedtoolcache/sccache/0.10.0/x64/sccache
    ACTIONS_CACHE_SERVICE_V2: on
    ACTIONS_RESULTS_URL: https://results-receiver.actions.githubusercontent.com/
    ACTIONS_RUNTIME_TOKEN: ***
    ```

File: rust_extension/src/config.rs [hunk 1]
```diff
@@ -125,14 +107,26 @@ impl LoggerConfigBuilder {
...
+    ///
+    /// This replaces any existing filters with the provided list.
+    pub fn with_filters<I, S>(mut self, filter_ids: I) -> Self
+    where
+        I: IntoIterator<Item = S>,
+        S: Into<String>,
+    {
+        self.filters = filter_ids.into_iter().map(Into::into).collect();
         self
     }
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/config.rs]: Has this now been resolved?

File: rust_extension/src/macros.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
+use pyo3::{prelude::*, PyResult};
+
+pub(crate) trait AsPyDict {
+    fn as_pydict(&self, py: Python<'_>) -> PyResult<PyObject>;
+}
+
+macro_rules! dict_set {
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/macros.rs]: Has this now been resolved?

File: rust_extension/src/macros.rs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    key: &str,
+    vec: &[T],
+) -> PyResult<()>
+where
+    T: IntoPyObject<'py> + Clone,
+{
+    if !vec.is_empty() {
+        let list = PyList::new(py, vec.iter().cloned())?;
+        dict.set_item(key, list)?;
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/macros.rs]: Has this now been resolved?

File: rust_extension/src/macros.rs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    opt: &Option<T>,
+) -> PyResult<()>
+where
+    T: IntoPyObject<'py> + Clone,
+{
+    if let Some(v) = opt {
+        dict.set_item(key, v.clone())?;
+    }
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/macros.rs]: Has this now been resolved?

File: rust_extension/src/macros.rs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    _py: Python<'py>,
+    dict: &Bound<'py, PyDict>,
+    key: &str,
+    val: &T,
+) -> PyResult<()>
+where
+    T: IntoPyObject<'py> + Clone,
+{
+    dict.set_item(key, val.clone())
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/macros.rs]: Has this now been resolved?

[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/macros.rs]: Has this now been resolved?

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review

File: rust_extension/src/macros.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/macros.rs]: @coderabbitai new GitHub issue, please

File: rust_extension/src/macros.rs [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    key: &str,
+    map: &BTreeMap<String, V>,
+) -> PyResult<()> {
+    if !map.is_empty() {
+        let sub = PyDict::new(py);
+        for (k, v) in map {
+            sub.set_item(k, v.as_pydict(py)?)?;
+        }
+        dict.set_item(key, sub)?;
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/macros.rs:80]: @coderabbitai new GitHub issue, please
